e2e-tests/src/test/treasury.rs

```diff
@@ -0,0 +1,199 @@
+    let issuance_after = get_total_issuance(&connection);
+    info!(
+        "[+] Treasury balance after tx: {}. Total issuance: {}.",
+        treasury_balance_after, issuance_after
+    );
+
+    assert!(
+        issuance_after <= issuance_before,
+        "Unexpectedly {} was minted",
+        issuance_after - issuance_before
+    );
+    assert!(
+        issuance_before <= issuance_after,
+        "Unexpectedly {} was burned",
+        issuance_before - issuance_after
+    );
+
+    let fee_info = get_tx_fee_info(&connection, &tx);
+    let fee = fee_info.fee_without_weight + fee_info.adjusted_weight;
+
+    assert_eq!(
+        treasury_balance_before + fee,
+        treasury_balance_after,
+        "Incorrect amount was channeled to the treasury: before = {}, after = {}, fee = {}",
+        treasury_balance_before,
+        treasury_balance_after,
+        fee
+    );
+
+    Ok(())
+}
+
+pub fn treasury_access(config: Config) -> anyhow::Result<()> {
+    let Config { node, seeds, .. } = config.clone();
+
+    let proposer = accounts_from_seeds(seeds)[0].to_owned();
+    let beneficiary = AccountId::from(proposer.public());
+    let connection = create_connection(node).set_signer(proposer);
+
+    propose_treasury_spend(0u128, &beneficiary, &connection);
+    propose_treasury_spend(0u128, &beneficiary, &connection);
+    let proposals_counter = get_proposals_counter(&connection);
+    assert!(proposals_counter >= 2, "Proposal was not created");
+
+    let sudo = get_sudo(config);
+    let connection = connection.set_signer(sudo);
+
+    treasury_approve(proposals_counter - 2, &connection)?;
+    treasury_reject(proposals_counter - 1, &connection)?;
+
+    Ok(())
+}
+
+fn get_total_issuance(connection: &Connection) -> u128 {
+    connection
+        .get_storage_value("Balances", "TotalIssuance", None)
+        .unwrap()
+        .unwrap()
+}
+
+fn get_treasury_account(connection: &Connection) -> AccountId32 {
+    let pallet_id = connection
+        .metadata
+        .module_with_constants_by_name("Treasury")
+        .unwrap()
+        .constant_by_name("PalletId")
+        .unwrap()
+        .get_value();
+    PalletId(pallet_id.try_into().unwrap()).into_account()
+}
+
+type ProposalTransaction =
+    UncheckedExtrinsicV4<([u8; 2], Compact<u128>, MultiAddress<AccountId, ()>)>;
+fn propose_treasury_spend(
+    value: u128,
+    beneficiary: &AccountId32,
+    connection: &Connection,
+) -> ProposalTransaction {
+    crate::send_extrinsic!(
+        connection,
+        "Treasury",
+        "propose_spend",
+        |tx_hash| info!("[+] Treasury spend transaction hash: {}", tx_hash),
+        Compact(value),
+        GenericAddress::Id(beneficiary.clone())
+    )
+}
+
+fn get_proposals_counter(connection: &Connection) -> u32 {
+    connection
+        .get_storage_value("Treasury", "ProposalCount", None)
+        .unwrap()
+        .unwrap()
+}
+
+type GovernanceTransaction = UncheckedExtrinsicV4<([u8; 2], Compact<u32>)>;
```

Comment:
to /utils/types.rs ?

Reply:
I have left this type (similarly to `ProposalTransaction`) as it is not returned from any public function and is for internal use only